docs/specs/adx/repo.md

```diff
@@ -0,0 +1,222 @@
+
+Some example schema IDs:
+
+```
+com.example.profile
+io.social.post
+net.users.bob.comment
+```
+
+## Schemas
+
+Schemas define the possible values of a record. Every record has a "type" which maps to a schema. Schemas are also used to distinguish collections of records, and are used to drive permissioning.
+
+### Schema distribution
+
+Schemas are designed to be machine-readable and network-accessible. While it is not current _required_ that a schema is available on the network, it is strongly advised to publish schemas so that a single canonical & authoritative representation is available to consumers of the method.
+
+To fetch a schema, a request must be sent to the fedrpc [`getSchema`](../fedrpc.md#getschema) method. This request is sent to the authority of the NSID.
+
+### Schema structure
+
+Record schemas are encoded in JSON and adhere to the following interface:
+
+```typescript
+interface RecordSchema {
+  adx: 1
+  id: string
+  revision?: number // a versioning counter
+  description?: string
+  record: JSONSchema
+}
+```
+
+Here is an example schema:
+
+```json
+{
+  "adx": 1,
+  "id": "com.example.post",
+  "schema": {
+    "type": "object",
+    "required": ["text", "createdAt"],
+    "properties": {
+      "text": {"type": "string", "maxLength": 256},
+      "createdAt": {"type": "string", "format": "date-time"}
+    }
+  }
+}
+```
+
+And here is a record using this example schema:
+
+```json
+{
+  "$type": "com.example.post",
+  "text": "Hello, world!",
+  "createdAt": "2022-09-15T16:37:17.131Z"
+}
+```
+
+### Reserved field names
+
+There are a set of fields which are reserved in ADX and shouldn't be used by schemas.
+
+|Field|Usage|
+|-|-|
+|`$type`|Declares the type of a record.|
+|`$ext`|Contains extensions to a record's base schema.|
+|`$required`|Used by extensions to flag whether their support is required.|
+|`$fallback`|Used by extensions to give a description of the missing data.|
+
+Generally it's wise to avoid `$` prefixes in your fieldnames.
+
+### Schema validation
+
+Constraints are structural: they apply constraints to fields under object path (eg `#/text`) to establish permissable values for that field. The constraints they can apply are value-type and valid values of the type (eg numbers within a range, strings of a certain format or pattern, etc). These constraints are described using [JSON Schema](https://json-schema.org/draft/2020-12/json-schema-core.html).
+
+Unconstrained fields are ignored during validation, but should be avoided in case future versions of the schema apply constraints.
+
+### Schema versioning
+
+Once a field constraint is published, it can never change. Loosening a constraint (adding possible values) will cause old software to fail validation for new data, and tightening a constraint (removing possible values) will cause new software to fail validation for old data. As a consequence, schemas may only add constraints to previously unconstrained fields.
```

Comment:
> schemas may only add constraints to previously unconstrained fields.

wouldn't this be "tightening a constraint?"

Reply:
Technically yes but we ignore unconstrained fields as a kind of special case, and if you do this then the new constraint has to be optional. I'll make sure that's noted.